planner/core/stats.go

```diff
@@ -150,12 +151,113 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
 			break
 		}
 	}
+	// consider the IndexMergePath.
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 2 && ds.ctx.GetSessionVars().EnableIndexMerge {
+		ds.generateIndexMergeOrPaths()
+	}
 	if ds.ctx.GetSessionVars().OptimizerSelectivityLevel >= 1 {
 		ds.stats.HistColl = finalHist
 	}
 	return ds.stats, nil
 }
 
+// getIndexMergeOrPath generates all possible IndexMergeOrPaths.
+func (ds *DataSource) generateIndexMergeOrPaths() {
+	usedIndexCount := len(ds.possibleAccessPaths) - 1
+	for i, cond := range ds.pushedDownConds {
+		var ixMergePartialIxPaths = make([]*accessPath, 0, 0)
+
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.LogicOr {
+			continue
+		}
+		dnfItems := expression.FlattenDNFConditions(sf)
+		for _, item := range dnfItems {
+			var indexAccessPaths []*accessPath
+			if itemF, ok := item.(*expression.ScalarFunction); ok && itemF.FuncName.L == ast.LogicAnd {
+				cnfItems := expression.FlattenCNFConditions(itemF)
+				indexAccessPaths = ds.buildAccessPath(cnfItems, usedIndexCount)
+			} else {
+				tempArgs := []expression.Expression{item}
+				indexAccessPaths = ds.buildAccessPath(tempArgs, usedIndexCount)
+			}
+			if indexAccessPaths == nil {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			ixMergePartialIxPath := ds.generateIndexMergePartialIndexPath(indexAccessPaths)
+			if ixMergePartialIxPath == nil {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			ixMergePartialIxPaths = append(ixMergePartialIxPaths, ixMergePartialIxPath)
+		}
+		if len(ixMergePartialIxPaths) > 1 {
+			possiblePath := ds.generateIndexMergeOrPath(ixMergePartialIxPaths, i)
+			if possiblePath != nil {
+				ds.possibleAccessPaths = append(ds.possibleAccessPaths, possiblePath)
+			}
+		}
+	}
+}
+
+// buildAccessPath generates all possible index paths for conditions.
+func (ds *DataSource) buildAccessPath(conditions []expression.Expression, usedIndexCount int) []*accessPath {
+	var results = make([]*accessPath, 0, 0)
+	for i := 1; i <= usedIndexCount; i++ {
+		newAccessPath := new(accessPath)
+		newAccessPath.index = ds.possibleAccessPaths[i].index
+		_, err := ds.deriveIndexPathStats(newAccessPath, conditions)
+		if err != nil {
+			return nil
+		}
+		// if accessConds is empty or tableFilter is not empty, we ignore the access path.
+		if len(newAccessPath.tableFilters) > 0 || len(newAccessPath.accessConds) == 0 {
```

Comment:
Condition `len(newAccessPath.tableFilters) > 0` seems too strict, we can carry it over to IndexMergeOr path actually? For simplicity, we can add this check now, please add a TODO here to remind optimizing this later.